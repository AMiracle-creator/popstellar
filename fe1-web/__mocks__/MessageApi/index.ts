export function requestCloseRollCall(
): Promise<void> {
  return new Promise((resolve) => {
    process.nextTick(() => resolve());
  });
}
