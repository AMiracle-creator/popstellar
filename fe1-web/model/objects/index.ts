import * as w from './wallet';

export const Wallet = w;

export * from './Base64Url';
export * from './Channel';
export * from './ConnectToLao';
export * from './Hash';
export * from './KeyPair';
export * from './Lao';
export * from './PopToken';
export * from './PrivateKey';
export * from './PublicKey';
export * from './Signature';
export * from './Timestamp';
export * from './WitnessSignature';
