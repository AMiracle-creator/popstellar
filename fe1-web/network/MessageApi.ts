import {
  EventTags, Hash, Lao, PublicKey, Timestamp,
} from 'model/objects';
import {
  CloseRollCall,
  CreateLao,
  CreateRollCall,
  OpenRollCall,
  ReopenRollCall,
  StateLao,
  UpdateLao,
  WitnessMessage,
} from 'model/network/method/message/data';
import {
  Channel, channelFromIds, ROOT_CHANNEL,
} from 'model/objects/Channel';
import {
  OpenedLaoStore, KeyPairStore,
} from 'store';
import { publish } from './JsonRpcApi';

/**
 * Adapts the starting time if start < creation.
 *
 * @param start
 * @param creation
 */
const adaptStartTime = (creation: Timestamp, start: Timestamp) => ((start.before(creation))
  ? creation : start);

/** Send a server query asking for the creation of a LAO with a given name (String) */
export function requestCreateLao(laoName: string): Promise<Channel> {
  const time = Timestamp.EpochNow();
  const pubKey = KeyPairStore.getPublicKey();

  const message = new CreateLao({
    id: Hash.fromStringArray(pubKey.toString(), time.toString(), laoName),
    name: laoName,
    creation: time,
    organizer: pubKey,
    witnesses: [],
  });

  return publish(ROOT_CHANNEL, message)
    .then(() => channelFromIds(message.id));
}

/** Send a server query asking for a LAO update providing a new name (String) */
export function requestUpdateLao(name: string, witnesses?: PublicKey[]): Promise<void> {
  const time: Timestamp = Timestamp.EpochNow();
  const currentLao: Lao = OpenedLaoStore.get();

  const message = new UpdateLao({
    id: Hash.fromStringArray(currentLao.organizer.toString(), currentLao.creation.toString(), name),
    name,
    last_modified: time,
    witnesses: (witnesses === undefined) ? currentLao.witnesses : witnesses,
  });

  return publish(channelFromIds(currentLao.id), message);
}

/** Send a server query asking for the current state of a LAO */
export function requestStateLao(): Promise<void> {
  const currentLao: Lao = OpenedLaoStore.get();

  const message = new StateLao({
    id: Hash.fromStringArray(
      currentLao.organizer.toString(), currentLao.creation.toString(), currentLao.name,
    ),
    name: currentLao.name,
    creation: currentLao.creation,
    last_modified: Timestamp.EpochNow(),
    organizer: currentLao.organizer,
    witnesses: currentLao.witnesses,
    modification_id: Hash.fromStringArray(), // FIXME need modification_id from storage
    modification_signatures: [], // FIXME need modification_signatures from storage
  });

  return publish(channelFromIds(currentLao.id), message);
}

/** Send a server message to acknowledge witnessing the message message (JS object) */
export function requestWitnessMessage(channel: Channel, messageId: Hash): Promise<void> {
  const message = new WitnessMessage({
    message_id: messageId,
    signature: KeyPairStore.getPrivateKey().sign(messageId),
  });

  return publish(channel, message);
}

/**
 * Send a server query asking for the creation of a roll call
 * @param name the name of the roll call event
 * @param location the location of the roll call event
 * @param proposedStart the time at which the attendee scanning will start (pending confirmation)
 * @param proposedEnd the time at which the attendee scanning will stop (pending confirmation)
 * @param description (optional) further information regarding the event
 * @return A Promise resolving when the operation is completed
 */
export function requestCreateRollCall(
  name: string, location: string,
  proposedStart: Timestamp, proposedEnd: Timestamp,
  description?: string,
): Promise<void> {
  const time: Timestamp = Timestamp.EpochNow();
  const currentLao: Lao = OpenedLaoStore.get();

  const message = new CreateRollCall({
    id: Hash.fromStringArray(
      EventTags.ROLL_CALL, currentLao.id.toString(), time.toString(), name,
    ),
    name: name,
    creation: time,
    location: location,
    proposed_start: adaptStartTime(time, proposedStart),
    proposed_end: proposedEnd,
    description: description,
  });

  const laoCh = channelFromIds(currentLao.id);
  return publish(laoCh, message);
}

/**
 * Send a server query asking for the opening of a roll call
 * @param rollCallId the ID of the roll call
 * @param start (optional) the time at which the operation happens, defaults to now.
 * @return A Promise resolving when the operation is completed
 */
export function requestOpenRollCall(rollCallId: Hash, start?: Timestamp): Promise<void> {
  const lao: Lao = OpenedLaoStore.get();
  const time = (start === undefined) ? Timestamp.EpochNow() : start;

  const message = new OpenRollCall({
    update_id: Hash.fromStringArray(
      EventTags.ROLL_CALL, lao.id.toString(), rollCallId.toString(), time.toString(),
    ),
    opens: rollCallId,
    opened_at: time,
  });

  return publish(channelFromIds(lao.id), message);
}

/**
 * Send a server query asking for the re-opening of a roll call
 * @param rollCallId the ID of the roll call
 * @param start (optional) the time at which the operation happens, defaults to now.
 * @return A Promise resolving when the operation is completed
 */
export function requestReopenRollCall(rollCallId: Hash, start?: Timestamp): Promise<void> {
  const lao: Lao = OpenedLaoStore.get();
  const time = (start === undefined) ? Timestamp.EpochNow() : start;

  const message = new ReopenRollCall({
    update_id: Hash.fromStringArray(
      EventTags.ROLL_CALL, lao.id.toString(), rollCallId.toString(), time.toString(),
    ),
    opens: rollCallId,
    opened_at: time,
  });

  return publish(channelFromIds(lao.id), message);
}

/**
 * Send a server query asking for the closing of a roll call
 * @param rollCallId the ID of the roll call
 * @param attendees list of the attendees' public keys
 * @param close (optional) the time at which the operation happens, defaults to now.
 * @return A Promise resolving when the operation is completed
 */
export function requestCloseRollCall(
  rollCallId: Hash,
  attendees: PublicKey[],
  close?: Timestamp,
): Promise<void> {
  const lao: Lao = OpenedLaoStore.get();
  const time = (close === undefined) ? Timestamp.EpochNow() : close;

  const message = new CloseRollCall({
    update_id: Hash.fromStringArray(
      EventTags.ROLL_CALL, lao.id.toString(), rollCallId.toString(), time.toString(),
    ),
    closes: rollCallId,
    closed_at: time,
    attendees: attendees,
  });

  return publish(channelFromIds(lao.id), message);
}
