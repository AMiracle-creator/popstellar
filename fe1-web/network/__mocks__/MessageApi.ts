export const requestCloseRollCall = jest.fn(() => Promise.resolve());
export const requestAddReaction = jest.fn(() => Promise.resolve());
