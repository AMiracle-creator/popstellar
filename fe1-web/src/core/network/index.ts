/**
 * The network module contains all the necessary code to communicate
 * with the backend systems, and abstracts away the intricacies of
 * network interactions from the rest of the application.
 *
 * This module specifically excludes the application-level parsing
 * of incoming messages, a task dealt with in the ingestion module.
 */

import { MessageRegistry } from './jsonrpc/messages';

export { getNetworkManager } from './NetworkManager';
export * from './NetworkError';
export * from 'core/network/RpcOperationError';

export function configure(messageRegistry: MessageRegistry) {}
