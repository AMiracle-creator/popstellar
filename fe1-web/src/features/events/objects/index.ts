export * from './Event';
export * from './LaoEventBuilder';
