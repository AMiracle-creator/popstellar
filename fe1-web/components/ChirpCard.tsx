import React, { useEffect, useState } from 'react';
import { View, Text } from 'react-native';
import PropTypes from 'prop-types';
import TimeAgo from 'react-timeago';
import { Ionicons } from '@expo/vector-icons';
import { Chirp } from 'model/objects/Chirp';
import DeleteButton from 'components/DeleteButton';
import styles from 'styles/stylesheets/chirpCard';
import { useSelector } from 'react-redux';
import { requestDeleteChirp } from '../network';
import { makeCurrentLao, makeEventGetter } from '../store';
import { PublicKey, RollCall } from '../model/objects';
import { generateToken } from '../model/objects/wallet';

/**
 * Component to display a chirp
 */
const ChirpCard = (props: IPropTypes) => {
  const { chirp } = props;
  const likesText = `  ${chirp.likes}`;

  const laoSelect = makeCurrentLao();
  const lao = useSelector(laoSelect);
  let userPublicKey: PublicKey | undefined;
  const [isSender, setIsSender] = useState(false);

  if (lao === undefined) {
    throw new Error('LAO is currently undefined, impossible to display chirps');
  }

  // This is temporary for now
  const zero = '  0';

  // Get the pop token of the user using the last tokenized roll call
  const rollCallId = lao.last_tokenized_roll_call_id;
  const eventSelect = makeEventGetter(lao.id, rollCallId);
  const rollCall: RollCall = useSelector(eventSelect) as RollCall;

  useEffect(() => {
    generateToken(lao.id, rollCallId).then((token) => {
      if (token && rollCall.containsToken(token)) {
        userPublicKey = token.publicKey;
        setIsSender(userPublicKey.valueOf() === chirp.sender.valueOf());
      }
    });
  }, [lao.last_tokenized_roll_call_id]);

  const deleteChirp = () => {
    if (userPublicKey) {
      requestDeleteChirp(userPublicKey, chirp.id)
        .catch((err) => {
          console.error('Could not remove chirp, error:', err);
        });
    } else {
      console.error('No token found for current user. '
        + 'Be sure to have participated in a Roll-Call.');
    }
  };

  return (
    <View style={styles.container}>
      <View style={styles.leftView}>
        <Ionicons name="person" size={40} color="black" />
      </View>
      <View style={styles.rightView}>
        <View style={styles.senderView}>
          <Text style={styles.senderText}>{chirp.sender.valueOf()}</Text>
        </View>
        <Text style={styles.chirpText}>{chirp.text}</Text>
        <View style={styles.reactionsView}>
          <View style={styles.reactionView}>
            <Ionicons name="thumbs-up" size={16} color="black" />
            <Text>{likesText}</Text>
          </View>
          <View style={styles.reactionView}>
            <Ionicons name="thumbs-down" size={16} color="black" />
            <Text>{zero}</Text>
          </View>
          <View style={styles.reactionView}>
            <Ionicons name="heart" size={16} color="black" />
            <Text>{zero}</Text>
          </View>
          <View style={styles.reactionView}>
            <Ionicons name="chatbubbles" size={16} color="black" />
            <Text>{zero}</Text>
          </View>
        </View>
        <View style={styles.timeView}>
          <TimeAgo date={chirp.time.valueOf() * 1000} />
        </View>
        { isSender && <DeleteButton action={() => { deleteChirp(); }} />}
      </View>
    </View>
  );
};

const propTypes = {
  chirp: PropTypes.instanceOf(Chirp).isRequired,
};

ChirpCard.prototype = propTypes;

type IPropTypes = {
  chirp: Chirp,
};

export default ChirpCard;
