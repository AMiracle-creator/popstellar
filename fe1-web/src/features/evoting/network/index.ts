import { ActionType, ObjectType } from 'core/network/jsonrpc/messages';
import { Hash } from 'core/objects';
import { dispatch, getStore } from 'core/redux';

import { EvotingConfiguration } from '../interface';
import { Election } from '../objects';
import { addElection, getElectionById, updateElection } from '../reducer';
import {
  handleCastVoteMessage,
  handleElectionEndMessage,
  handleElectionOpenMessage,
  handleElectionResultMessage,
  handleElectionSetupMessage,
} from './ElectionHandler';
import { CastVote, ElectionResult, EndElection, SetupElection } from './messages';
import { OpenElection } from './messages/OpenElection';

/**
 * Configures the network callbacks in a MessageRegistry.
 * @param configuration - An evoting config object
 */
export const configureNetwork = (configuration: EvotingConfiguration) => {
  // getElectionById bound to the global state
  const boundGetElectionById = (electionId: Hash | string) =>
    getElectionById(electionId, getStore().getState());

  const addElectionEvent = (laoId: Hash | string, election: Election) => {
    const electionState = election.toState();

    dispatch(addElection(electionState));
    dispatch(
      configuration.addEvent(laoId, {
        eventType: Election.EVENT_TYPE,
        id: electionState.id,
        start: electionState.start,
        end: electionState.start,
      }),
    );
  };

  const updateElectionEvent = (election: Election) => {
    const electionState = election.toState();

    dispatch(updateElection(electionState));
    dispatch(
      configuration.updateEvent({
        eventType: Election.EVENT_TYPE,
        id: electionState.id,
        start: electionState.start,
        end: electionState.start,
      }),
    );
  };

  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.SETUP,
    handleElectionSetupMessage(addElectionEvent),
    SetupElection.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.OPEN,
    handleElectionOpenMessage(boundGetElectionById, updateElectionEvent),
    OpenElection.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.CAST_VOTE,
    handleCastVoteMessage(boundGetElectionById, updateElectionEvent),
    CastVote.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.END,
    handleElectionEndMessage(boundGetElectionById, updateElectionEvent),
    EndElection.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.RESULT,
    handleElectionResultMessage(boundGetElectionById, updateElectionEvent),
    ElectionResult.fromJson,
  );
};
