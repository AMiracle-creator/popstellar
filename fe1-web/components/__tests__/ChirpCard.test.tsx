import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import {
  Hash, LaoState, PublicKey, Timestamp, Lao,
} from 'model/objects';
import { Chirp } from 'model/objects/Chirp';
import { requestDeleteChirp as mockRequestDeleteChirp } from 'network/MessageApi';
import { KeyPairStore, OpenedLaoStore } from 'store';
import ChirpCard from '../ChirpCard';

jest.mock('network/MessageApi');

const TIMESTAMP = 1609455600;
const laoState: LaoState = {
  id: '1234',
  name: 'MyLao',
  creation: TIMESTAMP,
  last_modified: TIMESTAMP,
  organizer: '1234',
  witnesses: [],
};

const sender = new PublicKey('Douglas Adams');
const chirp = new Chirp({
  id: new Hash('1234'),
  text: 'Don\'t panic.',
  sender: sender,
  time: new Timestamp(1609455600),
  likes: 42,
  isDeleted: false,
});

describe('ChirpCard', () => {
  it('renders correctly for sender', () => {
    const getMockLao = jest.spyOn(OpenedLaoStore, 'get');
    getMockLao.mockImplementation(() => Lao.fromState(laoState));
    const getMockSender = jest.spyOn(KeyPairStore, 'getPublicKey');
    getMockSender.mockImplementation(() => sender);
    const obj = render(
      <ChirpCard chirp={chirp} />,
    );
    expect(obj.toJSON()).toMatchSnapshot();
  });

  it('renders correctly for non-sender', () => {
    const getMockLao = jest.spyOn(OpenedLaoStore, 'get');
    getMockLao.mockImplementation(() => Lao.fromState(laoState));
    const getMockSender = jest.spyOn(KeyPairStore, 'getPublicKey');
    getMockSender.mockImplementation(() => new PublicKey('Marvin'));
    const obj = render(
      <ChirpCard chirp={chirp} />,
    );
    expect(obj.toJSON()).toMatchSnapshot();
  });

  it('calls delete correctly', () => {
    const getMockLao = jest.spyOn(OpenedLaoStore, 'get');
    getMockLao.mockImplementation(() => Lao.fromState(laoState));
    const getMockSender = jest.spyOn(KeyPairStore, 'getPublicKey');
    getMockSender.mockImplementation(() => sender);
    const button = render(
      <ChirpCard chirp={chirp} />,
    ).getByLabelText('delete');
    fireEvent.press(button);
    expect(mockRequestDeleteChirp).toHaveBeenCalledTimes(1);
  });
});
