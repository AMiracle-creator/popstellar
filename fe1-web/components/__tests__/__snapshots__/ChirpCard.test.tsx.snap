// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChirpCard render correct for a deleted chirp 1`] = `
<View
  style={
    Object {
      "borderColor": "gray",
      "borderTopWidth": 0,
      "borderWidth": 1,
      "flexDirection": "row",
      "padding": 10,
      "width": 600,
    }
  }
>
  <View
    style={
      Object {
        "width": 60,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "black",
            "fontSize": 40,
          },
          undefined,
          Object {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        Object {
          "fontSize": 18,
          "marginTop": 7,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
            "fontWeight": "600",
          }
        }
      >
        Douglas Adams
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "gray",
          "fontSize": 18,
          "paddingBottom": 20,
          "paddingTop": 10,
          "width": 520,
        }
      }
    >
      This chirp was deleted
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "fontSize": 18,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginRight": 10,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "black",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text>
            0
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignSelf": "flex-end",
          "marginTop": 10,
        }
      }
    >
      <time
        dateTime="2020-12-31T23:00:00.000Z"
        title="2020-12-31 23:00"
      >
        4 months ago
      </time>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Image
        accessibilityLabel="delete"
        source={1}
        style={
          Object {
            "height": 26,
            "width": 26,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`ChirpCard renders correctly for non-sender 1`] = `
<View
  style={
    Object {
      "borderColor": "gray",
      "borderTopWidth": 0,
      "borderWidth": 1,
      "flexDirection": "row",
      "padding": 10,
      "width": 600,
    }
  }
>
  <View
    style={
      Object {
        "width": 60,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "black",
            "fontSize": 40,
          },
          undefined,
          Object {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        Object {
          "fontSize": 18,
          "marginTop": 7,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
            "fontWeight": "600",
          }
        }
      >
        Douglas Adams
      </Text>
    </View>
    <Text
      style={
        Object {
          "fontSize": 18,
          "paddingBottom": 20,
          "paddingTop": 10,
          "width": 520,
        }
      }
    >
      Don't panic.
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "fontSize": 18,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginRight": 10,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "black",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text>
            0
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginRight": 10,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "black",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text>
            0
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginRight": 10,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "black",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text>
            0
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginRight": 10,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "black",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text>
            0
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignSelf": "flex-end",
          "marginTop": 10,
        }
      }
    >
      <time
        dateTime="2020-12-31T23:00:00.000Z"
        title="2020-12-31 23:00"
      >
        4 months ago
      </time>
    </View>
  </View>
</View>
`;

exports[`ChirpCard renders correctly for sender 1`] = `
<View
  style={
    Object {
      "borderColor": "gray",
      "borderTopWidth": 0,
      "borderWidth": 1,
      "flexDirection": "row",
      "padding": 10,
      "width": 600,
    }
  }
>
  <View
    style={
      Object {
        "width": 60,
      }
    }
  >
    <Text />
  </View>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        Object {
          "fontSize": 18,
          "marginTop": 7,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
            "fontWeight": "600",
          }
        }
      >
        Douglas Adams
      </Text>
    </View>
    <Text
      style={
        Object {
          "fontSize": 18,
          "paddingBottom": 20,
          "paddingTop": 10,
          "width": 520,
        }
      }
    >
      Don't panic.
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "fontSize": 18,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginRight": 10,
          }
        }
      >
        <Text />
        <Text>
            0
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginRight": 10,
          }
        }
      >
        <Text />
        <Text>
            0
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginRight": 10,
          }
        }
      >
        <Text />
        <Text>
            0
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginRight": 10,
          }
        }
      >
        <Text />
        <Text>
            0
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignSelf": "flex-end",
          "marginTop": 10,
        }
      }
    >
      <time
        dateTime="2020-12-31T23:00:00.000Z"
        title="2020-12-31 23:00"
      >
        4 months ago
      </time>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Image
        accessibilityLabel="delete"
        source={1}
        style={
          Object {
            "height": 26,
            "width": 26,
          }
        }
      />
    </View>
  </View>
</View>
`;
